hera_sim/utils.py

```diff
@@ -488,3 +498,286 @@ def _listify(x):
             return [x]
         else:
             return list(x)
+
+
+def reshape_vis(
+    vis: np.ndarray,
+    ant_1_array: np.ndarray,
+    ant_2_array: np.ndarray,
+    pol_array: np.ndarray,
+    antenna_numbers: np.ndarray,
+    n_times: int,
+    n_freqs: int,
+    n_ants: int,
+    n_pols: int,
+    invert: bool = False,
+    use_numba: bool = True,
+) -> np.ndarray:
+    """Reshaping helper for mutual coupling sims.
+
+    The mutual coupling simulations take as input, and return, a data array with
+    shape ``(Nblts, 1, Nfreqs, Npols)``, but perform matrix multiplications on
+    the data array reshaped to ``(Ntimes, Nfreqs, 2*Nants, 2*Nants)``. This
+    function performs the reshaping between the matrix multiply shape and the
+    input/output array shapes.
+
+    Parameters
+    ----------
+    vis
+        Input data array.
+    ant_1_array
+        Array specifying the first antenna in each baseline.
+    ant_2_array
+        Array specifying the second antenna in each baseline.
+    pol_array
+        Array specifying the observed polarizations via polarization numbers.
+    antenna_numbers
+        Array specifying all of the antennas to include in the reshaped data.
+    n_times
+        Number of integrations in the data.
+    n_freqs
+        Number of frequency channels in the data.
+    n_ants
+        Number of antennas.
+    n_pols
+        Number of polarizations in the data.
+    invert
+        Whether to reshape to :class:`pyuvdata.UVData`'s data array shape.
+    use_numba
+        Whether to use ``numba`` to speed up the reshaping.
+
+    Returns
+    -------
+    reshaped_vis
+        Input data reshaped to desired shape.
+    """
+    if invert:
+        out = np.zeros((ant_1_array.size, 1, n_freqs, n_pols), dtype=complex)
+    else:
+        out = np.zeros((n_times, n_freqs, 2 * n_ants, 2 * n_ants), dtype=complex)
+
+    # If we have numba, then this is a bit faster.
+    if HAVE_NUMBA and use_numba:
+        return jit_reshape_vis(
+            vis=vis,
+            out=out,
+            ant_1_array=ant_1_array,
+            ant_2_array=ant_2_array,
+            pol_array=pol_array,
+            antenna_numbers=antenna_numbers,
+            n_ants=n_ants,
+            n_pols=n_pols,
+            invert=invert,
+        )
+
+    # We don't have numba, so we need to do this a bit more slowly.
+    pol_slices = {"x": slice(None, None, 2), "y": slice(1, None, 2)}
+    for i, ai in enumerate(antenna_numbers):
+        for j, aj in enumerate(antenna_numbers[i:]):
+            j += i
+            for k, pol in enumerate(pol_array):
+                uvd_inds = np.argwhere(
+                    (ant_1_array == ai) & (ant_2_array == aj)
+                ).flatten()
+                p1, p2 = uvutils.polnum2str(pol)
```

Comment:
In my experience, this function call actually takes a bit of time. You could cache the result into a dict outside the big loops.

Reply:
I'm not sure I understand what you mean&mdash;the big loops only go over unique antenna pairs, so there's nothing to gain from caching locally. Moving the `uvd_inds = ...` snippet to outside the polarization loop would save some time when dealing with fully-polarized data, so I suppose I could make that change.

The only other thing I can think of would only save time on the second reshape (i.e. reshaping the crosstalk to match the data shape) by adding something like `uvd_ind_cache={}` to the function signature, but I'm not sure how I feel about that. I suppose it's fine if we're pretty confident that we're only going to be reshaping data for one `UVData` object per session, but I think that implementing a cache like this would cause some tests to break. Another option would be to require the user to provide the cache (i.e. instead doing `uvd_ind_cache=None`, then adding a `if uvd_ind_cache is None: uvd_ind_cache = {}` clause to the function body), but manually providing the cache feels a little annoying. (Also note that both of these options would only work for the non-jitted version of the code.)

Curious to hear what you think.

Reply:
I actually just meant the `polnum2str` function. You could do something like 

```python
polnumstr = {pol: uvutils.polnum2str(pol) for pol in pol_array}
    for i, ai in enumerate(antenna_numbers):
        for j, aj in enumerate(antenna_numbers[i:]):
            j += i
            for k, pol in enumerate(pol_array):
                p1, p2 = polnumstr[pol]
```

In your case it might not make much difference... I can't remember if it was when including the `x_orientation` that it was slower. But I definitely found that calling this for all baselines in H6C was a bottleneck.

Reply:
Ohh, gotcha. I didn't think that'd be a bottleneck, but that's a nice solution. I'll make the change. Thanks!